woocommerce/woocommerce
# Product Collection: Add validation for `__experimentalRegisterProductCollection` arguments

### Changes proposed in this Pull Request:





Closes #47963

This PR adds validation for arguments of the `registerProductCollection` function. Here are the highlights of this PR:
- It will print a warning in the console for any validation error. It's worth noting that it won't throw an error or bail early.
- It will ignore any extra properties that are not part of the API or aren't allowed. For example, `isDefault`, `isActive` etc. 
  - For this we are [constructing a new object](https://github.com/woocommerce/woocommerce/blob/ae267aa28ec83ab9f33309708d7c3aaf64857065/plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx#L339-L380) with only allowed properties.
- It will validate that all collection names should start with a prefix like [a-zA-Z0-9-]+/product-collection/[a-zA-Z0-9-]+. This will help in avoiding conflicts with other collections from different plugins. But it will just print a warning in the console if the collection name doesn't start with a prefix. It won't throw an error OR bail early.

These warnings will help 3PDs to debug their code and make sure that they are using the API correctly.



### How to test the changes in this Pull Request:



> [!IMPORTANT]
> To be tested by developers only.

1. Download following plugin, install and activate it: [register-product-collection-tester.zip](https://github.com/user-attachments/files/15633348/register-product-collection-tester.zip)
2. You will see that there are three collections registered in `collections.js` file using `__experimentalRegisterProductCollection` function:
   - My Custom Collection
   - My Custom Collection with Preview
   - My Custom Collection with Advanced Preview
3. Change various arguments passed to `__experimentalRegisterProductCollection` function in `collections.js` file and see if you are getting warning in console for validation errors, when you refresh the Editor page.
    - You can see all the validations in [this file](https://github.com/woocommerce/woocommerce/blob/ae267aa28ec83ab9f33309708d7c3aaf64857065/plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx). This will help you understand which arguments are validated & message you will see in console for each validation error.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Add validation for `__experimentalRegisterProductCollection` arguments

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
